### ISSUE TITLE
Model fine-tuning

### ISSUE THREAD

**Issue by majssssa** [usadellab/Helixer#209]
I attempted to fine-tune the land_plant model based on the annotation results of 22 species of Ranunculales. And a comparison was made using the land_plant and the fine-tuned model.

The result of the plant_model:

+confusion_matrix---------+----------+-----------+-------------+
|            | ig_pred    | utr_pred | exon_pred | intron_pred |
+------------+------------+----------+-----------+-------------+
| ig_ref     | 2642289327 | 14331613 | 78285881  | 77074733    |
| utr_ref    | 3936138    | 7377450  | 2458752   | 1498551     |
| exon_ref   | 1974244    | 371629   | 36559819  | 553334      |
| intron_ref | 8759673    | 1366656  | 2020805   | 39812901    |
+------------+------------+----------+-----------+-------------+

+normalized_confusion_matrix------+-----------+-------------+
|            | ig_pred | utr_pred | exon_pred | intron_pred |
+------------+---------+----------+-----------+-------------+
| ig_ref     | 0.9397  | 0.0051   | 0.0278    | 0.0274      |
| utr_ref    | 0.2578  | 0.4831   | 0.161     | 0.0981      |
| exon_ref   | 0.05    | 0.0094   | 0.9265    | 0.014       |
| intron_ref | 0.1686  | 0.0263   | 0.0389    | 0.7662      |
+------------+---------+----------+-----------+-------------+

+F1_summary--+---------+-----------+--------+----------+
|            | norm. H | Precision | Recall | F1-Score |
+------------+---------+-----------+--------+----------+
| ig         |         | 0.9945    | 0.9397 | 0.9663   |
| utr        |         | 0.3146    | 0.4831 | 0.3811   |
| exon       |         | 0.3064    | 0.9265 | 0.4605   |
| intron     |         | 0.3347    | 0.7662 | 0.4659   |
|            |         |           |        |          |
| legacy_cds |         | 0.3313    | 0.8636 | 0.4789   |
| sub_genic  |         | 0.3205    | 0.8354 | 0.4633   |
| genic      |         | 0.3200    | 0.7850 | 0.4547   |
+------------+---------+-----------+--------+----------+
Total acc: 0.9340

+confusion_matrix--------------+--------------+--------------+--------------+
|              | no_phase_pred | phase_0_pred | phase_1_pred | phase_2_pred |
+--------------+---------------+--------------+--------------+--------------+
| no_phase_ref | 2814926199    | 21461418     | 21457246     | 21367617     |
| phase_0_ref  | 1241747       | 11781591     | 58964        | 70595        |
| phase_1_ref  | 1248485       | 69005        | 11777542     | 58141        |
| phase_2_ref  | 1246364       | 57125        | 71044        | 11778423     |
+--------------+---------------+--------------+--------------+--------------+

+normalized_confusion_matrix---+--------------+--------------+--------------+
|              | no_phase_pred | phase_0_pred | phase_1_pred | phase_2_pred |
+--------------+---------------+--------------+--------------+--------------+
| no_phase_ref | 0.9777        | 0.0075       | 0.0075       | 0.0074       |
| phase_0_ref  | 0.0944        | 0.8957       | 0.0045       | 0.0054       |
| phase_1_ref  | 0.0949        | 0.0052       | 0.8954       | 0.0044       |
| phase_2_ref  | 0.0948        | 0.0043       | 0.0054       | 0.8955       |
+--------------+---------------+--------------+--------------+--------------+

+F1_summary+---------+-----------+--------+----------+
|          | norm. H | Precision | Recall | F1-Score |
+----------+---------+-----------+--------+----------+
| no_phase |         | 0.9987    | 0.9777 | 0.9881   |
| phase_0  |         | 0.3531    | 0.8957 | 0.5065   |
| phase_1  |         | 0.3530    | 0.8954 | 0.5064   |
| phase_2  |         | 0.3540    | 0.8955 | 0.5074   |
+----------+---------+-----------+--------+----------+

The result of the fine-tuning of mode:

+confusion_matrix---------+----------+-----------+-------------+
|            | ig_pred    | utr_pred | exon_pred | intron_pred |
+------------+------------+----------+-----------+-------------+
| ig_ref     | 2797285941 | 1107563  | 5371550   | 8216500     |
| utr_ref    | 9106931    | 3906697  | 1499652   | 757611      |
| exon_ref   | 12042365   | 137219   | 26831949  | 447493      |
| intron_ref | 21153034   | 381220   | 720899    | 29704882    |
+------------+------------+----------+-----------+-------------+

+normalized_confusion_matrix------+-----------+-------------+
|            | ig_pred | utr_pred | exon_pred | intron_pred |
+------------+---------+----------+-----------+-------------+
| ig_ref     | 0.9948  | 0.0004   | 0.0019    | 0.0029      |
| utr_ref    | 0.5964  | 0.2558   | 0.0982    | 0.0496      |
| exon_ref   | 0.3052  | 0.0035   | 0.68      | 0.0113      |
| intron_ref | 0.4071  | 0.0073   | 0.0139    | 0.5717      |
+------------+---------+----------+-----------+-------------+

+F1_summary--+---------+-----------+--------+----------+
|            | norm. H | Precision | Recall | F1-Score |
+------------+---------+-----------+--------+----------+
| ig         |         | 0.9851    | 0.9948 | 0.9899   |
| utr        |         | 0.7061    | 0.2558 | 0.3756   |
| exon       |         | 0.7795    | 0.6800 | 0.7263   |
| intron     |         | 0.7592    | 0.5717 | 0.6522   |
|            |         |           |        |          |
| legacy_cds |         | 0.7846    | 0.6312 | 0.6996   |
| sub_genic  |         | 0.7687    | 0.6184 | 0.6854   |
| genic      |         | 0.7643    | 0.5665 | 0.6507   |
+------------+---------+-----------+--------+----------+
Total acc: 0.9791

+confusion_matrix--------------+--------------+--------------+--------------+
|              | no_phase_pred | phase_0_pred | phase_1_pred | phase_2_pred |
+--------------+---------------+--------------+--------------+--------------+
| no_phase_ref | 2873276528    | 1982772      | 1974735      | 1978445      |
| phase_0_ref  | 4911091       | 8177096      | 25377        | 39333        |
| phase_1_ref  | 4917424       | 39832        | 8169504      | 26413        |
| phase_2_ref  | 4909163       | 26081        | 39368        | 8178344      |
+--------------+---------------+--------------+--------------+--------------+

+normalized_confusion_matrix---+--------------+--------------+--------------+
|              | no_phase_pred | phase_0_pred | phase_1_pred | phase_2_pred |
+--------------+---------------+--------------+--------------+--------------+
| no_phase_ref | 0.9979        | 0.0007       | 0.0007       | 0.0007       |
| phase_0_ref  | 0.3734        | 0.6217       | 0.0019       | 0.003        |
| phase_1_ref  | 0.3739        | 0.003        | 0.6211       | 0.002        |
| phase_2_ref  | 0.3732        | 0.002        | 0.003        | 0.6218       |
+--------------+---------------+--------------+--------------+--------------+

+F1_summary+---------+-----------+--------+----------+
|          | norm. H | Precision | Recall | F1-Score |
+----------+---------+-----------+--------+----------+
| no_phase |         | 0.9949    | 0.9979 | 0.9964   |
| phase_0  |         | 0.7997    | 0.6217 | 0.6995   |
| phase_1  |         | 0.8002    | 0.6211 | 0.6994   |
| phase_2  |         | 0.8000    | 0.6218 | 0.6997   |
+----------+---------+-----------+--------+----------+
Total acc: 0.9928

Although the total acc has improved, strangely enough, the recall rate has decreased significantly while the precision has increased substantially. In other words, the results predicted by the fine-tuned model are fewer but more precision. I would like to know if such a minor adjustment leads to a positive outcome?

**Comment by felicitas215** [comment]
Hi, thank you for your question. It looks like your model improved on predicting the intergenic region correctly, but got worse for UTR, CDS and intron (see the first normalized confusion matrix). Could you tell me what exact command you were using for fine-tuning? These results look like the class weights might be skewed.

PS: The total accuracy is not indicative of how well the model is doing, since it favors the most abundant class (intergenic).

**Comment by majssssa** [comment]
Thank you for your recovery.

`apptainer run --env PYTHONNOUSERSITE=1 --nv ~/helixer-docker_latest.sif HybridModel.py --data-dir ./h5 --save-model-path best_helixer_model.h5 --epochs 50 --predict-phase --load-model-path /home/majunpeng/.local/share/Helixer/models/land_plant/land_plant_v0.3_a_0080.h5 --resume-training --batch-size 80 --val-test-batch-size 160`

This is the command for fine-tuning the model. It seems that I haven't set any parameters related to weights.

**Comment by majssssa** [comment]
Thank you for your response. I have added these two parameters and re-performed the model fine-tuning.
---- Replied Message ----
FromFelicitas ***@***.***>Date02/04/2026 21:56 ***@***.***> CcJunpeng ***@***.***>,
***@***.***>SubjectRe: [usadellab/Helixer] Model fine-tuning (Issue #209)
felicitas215 left a comment (usadellab/Helixer#209)
Yes, I'm sorry we don't make it clear that class and transition weights should be used during fine-tuning. This will be fixed soon. The standard weights for training are --class-weights "[0.7, 1.6, 1.2, 1.2]" --transition-weights "[1, 12, 3, 1, 12, 3]". Adding this should give you a better result. Please let me know if you have any further questions.

**Comment by felicitas215** [comment]
I will close this issue - if your question is not sufficiently answered, please reopen the issue.
